File: references/dd/tools/check-dd-reference-fitness.py
#!/usr/bin/env python3
"""
DD reference metadata fitness checks.

Runs the principles from dd-reference-fitness-principles.md against one or more dd-{ver}.json
payloads. Exits non-zero on any failure. Port of the former check-dd-reference-fitness.js.

Usage:
  python3 check-dd-reference-fitness.py <dd-json-path> [<dd-json-path> ...]

Defaults to scanning references/dd/json/ if no paths are supplied.
"""

from __future__ import annotations

import json
import re
import sys
from pathlib import Path


def is_edm(t) -> bool:
    return isinstance(t, str) and t.startswith("Edm.")


def non_empty_string(v) -> bool:
    return isinstance(v, str) and len(v) > 0


def version_from_filename(path: str):
    m = re.match(r"^dd-(.+)\.json$", Path(path).name)
    return m.group(1) if m else None


# ── Per-file checks: each returns {"ok": bool, "detail"?: str, "info"?: str} ──


def check_resource_completeness(d, _ctx):
    resources = set(d.get("resources") or [])
    missing = [r for r in {f["resourceName"] for f in d["fields"]} if r not in resources]
    if not missing:
        return {"ok": True}
    return {"ok": False, "detail": f"missing from resources[]: {', '.join(missing)}"}


def check_expansion_shape(d, _ctx):
    exp = [f for f in d["fields"] if f.get("isExpansion") is True]
    bad = [
        f for f in exp
        if not non_empty_string(f.get("type"))
        or not non_empty_string(f.get("typeName"))
        or not non_empty_string(f.get("sourceResource"))
    ]
    if not bad:
        return {"ok": True, "info": f"{len(exp)} expansions checked"}
    listed = "\n  ".join(
        f"{f['resourceName']}.{f['fieldName']} "
        f"(type={f.get('type')}, typeName={f.get('typeName')}, sourceResource={f.get('sourceResource')})"
        for f in bad[:10]
    )
    return {"ok": False, "detail": f"{len(bad)} malformed:\n  {listed}"}


def check_sourceresource_resolves(d, _ctx):
    resources = set(d.get("resources") or [])
    exp = [f for f in d["fields"] if f.get("isExpansion") is True]
    unresolved = [
        f"{f['resourceName']}.{f['fieldName']} → {f['sourceResource']}"
        for f in exp
        if non_empty_string(f.get("sourceResource")) and f["sourceResource"] not in resources
    ]
    if not unresolved:
        return {"ok": True}
    return {"ok": False, "detail": f"{len(unresolved)} unresolved:\n  " + "\n  ".join(unresolved[:10])}


def check_non_edm_resolves(d, _ctx):
    lookup_names = {l_["lookupName"] for l_ in (d.get("lookups") or [])}
    candidates = [
        f for f in d["fields"]
        if not is_edm(f.get("type")) and not f.get("isExpansion") and f.get("lookupStatus") != "Open"
    ]
    unresolved = [
        f"{f['resourceName']}.{f['fieldName']} → {f['type']} (lookupStatus={f.get('lookupStatus', 'absent')})"
        for f in candidates
        if f.get("type") not in lookup_names
    ]
    if not unresolved:
        return {"ok": True, "info": f"{len(candidates)} non-Edm fields checked"}
    return {"ok": False, "detail": f"{len(unresolved)} unresolved:\n  " + "\n  ".join(unresolved[:10])}


def check_lookup_completeness(d, _ctx):
    lookups = d.get("lookups") or []
    bad = [
        l_ for l_ in lookups
        if not non_empty_string(l_.get("lookupName")) or not non_empty_string(l_.get("lookupValue"))
    ]
    if not bad:
        return {"ok": True, "info": f"{len(lookups)} lookups checked"}
    return {"ok": False, "detail": f"{len(bad)} malformed (first: {json.dumps(bad[0])})"}


def check_version_header_match(d, ctx):
    expected = version_from_filename(ctx["filename"])
    if not expected:
        return {"ok": True, "info": "filename has no version segment; skipping"}
    if d.get("version") == expected:
        return {"ok": True}
    return {"ok": False, "detail": f"filename says {expected}, top-level version is {d.get('version')}"}


CHECKS = [
    ("resource-completeness", "every field.resourceName appears in resources[]", check_resource_completeness),
    ("expansion-shape", "every expansion has type, typeName, sourceResource", check_expansion_shape),
    ("sourceresource-resolves", "every expansion sourceResource resolves to a resource", check_sourceresource_resolves),
    ("non-edm-resolves", 'every non-Edm field type (other than lookupStatus="Open") resolves to a lookup', check_non_edm_resolves),
    ("lookup-completeness", "every lookup has non-empty lookupName and lookupValue", check_lookup_completeness),
    ("version-header-match", "top-level version matches filename", check_version_header_match),
]

DEFAULT_PATHS = [
    "references/dd/json/dd-1.7.json",
    "references/dd/json/dd-2.0.json",
    "references/dd/json/dd-2.1.json",
]


def main() -> None:
    arg_paths = sys.argv[1:]
    candidates = arg_paths if arg_paths else DEFAULT_PATHS
    # absolute(), not resolve(): match Node path.resolve (no symlink following) for identical output.
    paths = [str(Path(p).absolute()) for p in candidates if Path(p).exists()]

    if not paths:
        print("No JSON files found.", file=sys.stderr)
        sys.exit(2)

    total_fail = 0
    files_by_version: dict = {}

    for path in paths:
        data = json.loads(Path(path).read_text(encoding="utf-8"))
        ver = version_from_filename(path)
        files_by_version.setdefault(ver, []).append({"path": path, "data": data})

        print(f"\n── {path} ──")
        for name, description, run in CHECKS:
            result = run(data, {"filename": path})
            if result["ok"]:
                info = f" ({result['info']})" if result.get("info") else ""
                print(f"  ✓ {name}: pass{info}")
            else:
                print(f"  ✗ {name}: {description}")
                print(f"    {result['detail']}")
                total_fail += 1

    # Cross-version FK consistency — runs across versions when >=2 are present.
    print("\n── cross-version checks ──")
    all_versions = [
        {"version": v, "data": files[0]["data"]}
        for v, files in files_by_version.items() if v
    ]
    if len(all_versions) >= 2:
        fk_by_key: dict = {}
        for entry in all_versions:
            for f in entry["data"]["fields"]:
                if f.get("isExpansion") and non_empty_string(f.get("sourceResource")):
                    key = f"{f['resourceName']}.{f['fieldName']}"
                    fk_by_key.setdefault(key, []).append(
                        {"version": entry["version"], "sourceResource": f["sourceResource"]}
                    )
        drift = []
        for key, entries in fk_by_key.items():
            targets = {e["sourceResource"] for e in entries}
            if len(targets) > 1:
                joined = ", ".join(f"{e['version']}→{e['sourceResource']}" for e in entries)
                drift.append(f"{key}: {joined}")
        if not drift:
            print(f"  ✓ cross-version-fk-consistency: pass ({len(fk_by_key)} FK fields shared across versions)")
        else:
            print(f"  ⚠ cross-version-fk-consistency: {len(drift)} fields drift across versions")
            for d_ in drift[:10]:
                print(f"    {d_}")
            print("    (note: cross-version drift is an audit-flag, not a fail — verify against version rename docs)")
    else:
        print("  (skipped — need ≥2 versions for cross-version check)")

    print()
    if total_fail == 0:
        print(f"✓ All checks passed across {len(paths)} file(s).")
        sys.exit(0)
    else:
        print(f"✗ {total_fail} check failure(s) across {len(paths)} file(s).")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: references/dd/tools/check-dd-reference-fitness.js
#!/usr/bin/env node
/**
 * DD reference metadata fitness checks.
 *
 * Runs the principles from reso-certification/docs/dd-reference-fitness-principles.md
 * against one or more dd-{ver}.json payloads. Exits non-zero on any failure.
 *
 * Usage:
 *   node check-dd-reference-fitness.js <dd-json-path> [<dd-json-path> ...]
 *
 * Defaults to scanning both reference-metadata/ and src/etl/reference-metadata/
 * if no paths are supplied.
 */

import { readFileSync, existsSync } from 'node:fs';
import { resolve, basename } from 'node:path';

const isEdm = (t) => typeof t === 'string' && t.startsWith('Edm.');
const nonEmptyString = (v) => typeof v === 'string' && v.length > 0;

const versionFromFilename = (path) => {
  const m = basename(path).match(/^dd-(.+)\.json$/);
  return m ? m[1] : null;
};

const checks = [
  // 2. Resource list completeness vs fields
  {
    name: 'resource-completeness',
    description: 'every field.resourceName appears in resources[]',
    run: (d) => {
      const resources = new Set(d.resources ?? []);
      const missing = [...new Set(d.fields.map((f) => f.resourceName))]
        .filter((r) => !resources.has(r));
      return missing.length === 0
        ? { ok: true }
        : { ok: false, detail: `missing from resources[]: ${missing.join(', ')}` };
    },
  },
  // 3. Expansion shape
  {
    name: 'expansion-shape',
    description: 'every expansion has type, typeName, sourceResource',
    run: (d) => {
      const exp = d.fields.filter((f) => f.isExpansion === true);
      const bad = exp.filter(
        (f) =>
          !nonEmptyString(f.type) ||
          !nonEmptyString(f.typeName) ||
          !nonEmptyString(f.sourceResource),
      );
      return bad.length === 0
        ? { ok: true, info: `${exp.length} expansions checked` }
        : {
            ok: false,
            detail:
              `${bad.length} malformed:\n  ` +
              bad
                .slice(0, 10)
                .map(
                  (f) =>
                    `${f.resourceName}.${f.fieldName} (type=${f.type}, typeName=${f.typeName}, sourceResource=${f.sourceResource})`,
                )
                .join('\n  '),
          };
    },
  },
  // 4. sourceResource resolves to known resource
  {
    name: 'sourceresource-resolves',
    description: 'every expansion sourceResource resolves to a resource',
    run: (d) => {
      const resources = new Set(d.resources ?? []);
      const exp = d.fields.filter((f) => f.isExpansion === true);
      const unresolved = exp
        .filter(
          (f) => nonEmptyString(f.sourceResource) && !resources.has(f.sourceResource),
        )
        .map((f) => `${f.resourceName}.${f.fieldName} → ${f.sourceResource}`);
      return unresolved.length === 0
        ? { ok: true }
        : { ok: false, detail: `${unresolved.length} unresolved:\n  ` + unresolved.slice(0, 10).join('\n  ') };
    },
  },
  // 5. Non-Edm types resolve to lookups (excluding expansions and lookupStatus=Open)
  {
    name: 'non-edm-resolves',
    description: 'every non-Edm field type (other than lookupStatus="Open") resolves to a lookup',
    run: (d) => {
      const lookupNames = new Set((d.lookups ?? []).map((l) => l.lookupName));
      const candidates = d.fields.filter(
        (f) =>
          !isEdm(f.type) && !f.isExpansion && f.lookupStatus !== 'Open',
      );
      const unresolved = candidates
        .filter((f) => !lookupNames.has(f.type))
        .map((f) => `${f.resourceName}.${f.fieldName} → ${f.type} (lookupStatus=${f.lookupStatus ?? 'absent'})`);
      return unresolved.length === 0
        ? { ok: true, info: `${candidates.length} non-Edm fields checked` }
        : { ok: false, detail: `${unresolved.length} unresolved:\n  ` + unresolved.slice(0, 10).join('\n  ') };
    },
  },
  // 6. Lookup completeness
  {
    name: 'lookup-completeness',
    description: 'every lookup has non-empty lookupName and lookupValue',
    run: (d) => {
      const bad = (d.lookups ?? []).filter(
        (l) => !nonEmptyString(l.lookupName) || !nonEmptyString(l.lookupValue),
      );
      return bad.length === 0
        ? { ok: true, info: `${d.lookups?.length ?? 0} lookups checked` }
        : { ok: false, detail: `${bad.length} malformed (first: ${JSON.stringify(bad[0])})` };
    },
  },
  // 10. Version header matches filename
  {
    name: 'version-header-match',
    description: 'top-level version matches filename',
    run: (d, { filename }) => {
      const expected = versionFromFilename(filename);
      if (!expected) return { ok: true, info: 'filename has no version segment; skipping' };
      return d.version === expected
        ? { ok: true }
        : { ok: false, detail: `filename says ${expected}, top-level version is ${d.version}` };
    },
  },
];

const DEFAULT_PATHS = [
  'reso-certification/reference-metadata/dd-1.7.json',
  'reso-certification/reference-metadata/dd-2.0.json',
  'reso-certification/reference-metadata/dd-2.1.json',
  'reso-certification/src/etl/reference-metadata/dd-1.7.json',
  'reso-certification/src/etl/reference-metadata/dd-2.0.json',
  'reso-certification/src/etl/reference-metadata/dd-2.1.json',
];

const argPaths = process.argv.slice(2);
const paths = (argPaths.length > 0 ? argPaths : DEFAULT_PATHS)
  .map((p) => resolve(p))
  .filter((p) => existsSync(p));

if (paths.length === 0) {
  console.error('No JSON files found.');
  process.exit(2);
}

let totalFail = 0;
const filesByVersion = new Map();

for (const path of paths) {
  const data = JSON.parse(readFileSync(path, 'utf-8'));
  const ver = versionFromFilename(path);
  if (!filesByVersion.has(ver)) filesByVersion.set(ver, []);
  filesByVersion.get(ver).push({ path, data });

  console.log(`\n── ${path} ──`);
  for (const check of checks) {
    const result = check.run(data, { filename: path });
    if (result.ok) {
      console.log(`  ✓ ${check.name}: pass${result.info ? ` (${result.info})` : ''}`);
    } else {
      console.log(`  ✗ ${check.name}: ${check.description}`);
      console.log(`    ${result.detail}`);
      totalFail++;
    }
  }
}

// 7. Cross-version FK consistency — runs across versions when ≥2 are present
console.log('\n── cross-version checks ──');
const allVersions = [...filesByVersion.entries()]
  .filter(([v]) => v)
  .map(([v, files]) => ({ version: v, data: files[0].data }));

if (allVersions.length >= 2) {
  const fkByKey = new Map();
  for (const { version, data } of allVersions) {
    for (const f of data.fields.filter((f) => f.isExpansion && nonEmptyString(f.sourceResource))) {
      const key = `${f.resourceName}.${f.fieldName}`;
      if (!fkByKey.has(key)) fkByKey.set(key, []);
      fkByKey.get(key).push({ version, sourceResource: f.sourceResource });
    }
  }
  const drift = [];
  for (const [key, entries] of fkByKey) {
    const targets = new Set(entries.map((e) => e.sourceResource));
    if (targets.size > 1) {
      drift.push(
        `${key}: ${entries.map((e) => `${e.version}→${e.sourceResource}`).join(', ')}`,
      );
    }
  }
  if (drift.length === 0) {
    console.log(`  ✓ cross-version-fk-consistency: pass (${fkByKey.size} FK fields shared across versions)`);
  } else {
    console.log(`  ⚠ cross-version-fk-consistency: ${drift.length} fields drift across versions`);
    drift.slice(0, 10).forEach((d) => console.log(`    ${d}`));
    console.log('    (note: cross-version drift is an audit-flag, not a fail — verify against version rename docs)');
  }
} else {
  console.log('  (skipped — need ≥2 versions for cross-version check)');
}

console.log();
if (totalFail === 0) {
  console.log(`✓ All checks passed across ${paths.length} file(s).`);
  process.exit(0);
} else {
  console.log(`✗ ${totalFail} check failure(s) across ${paths.length} file(s).`);
  process.exit(1);
}
